docs/zig-and-c.org

```diff
@@ -0,0 +1,249 @@
+#+begin_src zig
+pub const c = @cImport({
+    @cInclude("stdio.h");
+    @cInclude("stdlib.h");
+    @cInclude("image.h");
+});
+#+end_src
+Then call this zig file in your other zig files.
+
+*** How to call a Zig function from C / Continuing a C project with Zig
+You can continue your C project without using Clang or GCC but by using Zig with all its toolchain it comes with.
+
+In order to have a c file (main.c) as the entry point of your project using the zig build tool you have to add/comment the following lines to your build.zig file:
+#+begin_src zig
+const exe = b.addExecutable(.{
+    .name = "c_project",
+    // .root_source_file = b.path("src/main.zig"),
+    .target = target,
+    .optimize = optimize,
+});
+exe.root_module.addCSourceFile(.{ .file = .{ .path = "src/main.c" }, .flags = &.{"-std=c99"} });
+exe.linkLibC();
+#+end_src
+
+If you want to have more C files than just main.c you can add them like so:
+#+begin_src zig
+exe.addCSourceFile(.{ .file = .{ .path = "c-src/image.c" }, .flags = &.{"-std=c99"} });
+#+end_src
+
+TODO pros and cons of using zig toolchain instead of clang or gcc
+
+*** How is it done under the hood
+When you do @cImport(@cInclude("foo.h")) in your zig code it runs translate-c and exposes the function and type definitions.
+
+TODO talk about C ABI
+
+*** Util to translate C code to Zig
+zig translate-c is an util built in the zig toolchains that allows you to translate C code to Zig code.
+You can translate any code but the code is going to be UNREADABLE, so I would not recommend this tool if you plan on modifying the code afterwards.
+You have better time importing the C code in your Zig code.
+Note that if you want to translate a C file that uses the libc you have to add the -lc flag like so:
+#+begin_src shell
+zig translate-c main.c -lc
+#+end_src
+
+**** Comparison with other langauges that use C code
+To test if integrating C code in Zig projects is really as seemless as some claims, I have decided to compare the C integration with Python aswell.
+
+In order to do that I wrote a small C library:
+#+begin_src c
+int add(int a, int b) { return a + b; }
+#+end_src
+
+What I am going to do is test how much time it takes each program to run this function x times. (in this case x = 100'000'000)
+
+Note: I didnt use any optimization in flag in python (because it did not change anything) and neither in Zig because I didnt want the compiler to try to optimize the code and be lazy.
+
+Then in order to compare the 2 languages I wrote 4 programs:
+- Zig code that has and add function implementation in Zig
+#+begin_src zig
+const std = @import("std");
+
+fn add(a: u32, b: u32) u32 {
+    return a + b;
+}
+
+pub fn main() !void {
+    var i: usize = 0;
+    while (i < 100000000) : (i += 1) {
+        _ = add(3, 7);
+    }
+    std.debug.print("done\n", .{});
+}
+#+end_src
+Result: ~0.38sec
```

Comment:
You run the code only once ? Maybe use hyperfine : https://github.com/sharkdp/hyperfine

Reply:
no, ran it multiple times
im going to use it hyperfine next time :)